src/tokenserver/extractors.rs

```diff
@@ -26,34 +33,74 @@ impl FromRequest for TokenserverRequest {
     type Error = Error;
     type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;
 
-    fn from_request(req: &HttpRequest, payload: &mut Payload) -> Self::Future {
+    fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
         let req = req.clone();
-        let mut payload = payload.take();
 
         Box::pin(async move {
-            let token_data = TokenData::from_request(&req, &mut payload).await?;
+            let token_data = TokenData::extract(&req).await?;
+            let state = get_server_state(&req)?;
+            let tokenserver_state = state.tokenserver_state.as_ref().unwrap();
+            let fxa_metrics_hash_secret = &tokenserver_state.fxa_metrics_hash_secret.as_bytes();
+            let shared_secret =
+                String::from_utf8(state.secrets.master_secret.clone()).map_err(|_| {
+                    error!("⚠️ Failed to read master secret");
+
+                    TokenserverError::internal_error()
+                })?;
+            let key_id = KeyId::extract(&req).await?;
+            let fxa_uid = token_data.user;
+            let hashed_fxa_uid = {
+                let hashed_fxa_uid_full = fxa_metrics_hash(&fxa_uid, fxa_metrics_hash_secret);
+                hashed_fxa_uid_full[0..32].to_owned()
+            };
+            let hashed_device_id = {
+                let device_id = "none".to_string();
+                hash_device_id(&hashed_fxa_uid, &device_id, fxa_metrics_hash_secret)
+            };
             let service_id = {
                 let path = req.match_info();
 
-                match (path.get("application"), path.get("version")) {
-                    (Some("sync"), Some("1.1")) => db::SYNC_1_1_SERVICE_ID,
-                    (Some("sync"), Some("1.5")) => db::SYNC_1_5_SERVICE_ID,
-                    // XXX: This error will be replaced with a more descriptive error as part of
-                    // #1133
-                    _ => {
-                        return Err(ValidationErrorKind::FromDetails(
-                            "Invalid application and version".to_owned(),
-                            RequestErrorLocation::Path,
-                            None,
-                            None,
+                // If we've reached this extractor, we know that the Tokenserver path was matched,
+                // meaning "application" and "version" are both present in the URL. So, we can use
+                // `unwrap()` here.
+                let application = path.get("application").unwrap();
+                let version = path.get("version").unwrap();
+
+                if application == "sync" {
```

Comment:
This will likely eventually be removed, since we're likely going to get rid of the `services` table